nservicebus/upgrades/asp-saga-deduplication.md

```diff
@@ -0,0 +1,78 @@
+
+
+## Patch steps
+
+Steps:
+1. Run de-duplication utility
+2. Update NServiceBus.Azure dependency in all endpoints that use it
+3. Run de-duplication utility
+
+
+NOTE: Minimizing the time between steps one and two will minimize the work required later in the process.
+
+### Upgrade endpoint configurations
+Update references to NServiceBus.Azure to version 6.2.4 or later.
+
+### Running the de-duplication utility
+1. Download the de-duplication tool from xxxxxxxxx
+2. Run the de-duplication tool from any computer that has both an internet connection and the .NET Framework v4.5.2 installed
+3. Open a commandline and run the following command: `NServiceBus.AzureStoragePersistence.SagaDeduplicator.exe directory=<directory> sagaTypeName=<sagaTypeName> sagaProperty=<sagaProperty> operation=Download connectionString=<connectionString>`
+	- **directory**: This should be the full path of the working directory that the de-duplication tool will use. The tool will need to have read, write and delete permissions on the folder selected. If there are any spaces in the directory path this value should be enclosed in double quotes. Example: `directory="C:\my working folder\deduplication"`
+	- **sagaTypeName**:
+	- **sagaProperty**:
+	- **operation**: "Download"
+	- **connectionString**: This parameter is optional and there are two ways to provide the Azure connection string.
+		1. As a commandline parameter. Example: `connectionString=DefaultEndpointsProtocol=https;AccountName=anAzureStorageAccount;AccountKey=FuiDeGgJoHackI4h1d44LFK0ceqfkxDMO79In+ofWG42rw9kNDyQuErNmJ8WyJ7GLa+xy5NVaN21lOMZUevaQw==;BlobEndpoint=https://anAzureStorageAccount.blob.core.windows.net/;TableEndpoint=https://anAzureStorageAccount.table.core.windows.net/;QueueEndpoint=https://anAzureStorageAccount.queue.core.windows.net/;FileEndpoint=https://anAzureStorageAccount.file.core.windows.net/`
+		2. As a connectionStrings entry in the `NServiceBus.AzureStoragePersistence.SagaDeduplicator.exe.config`. Example:  
+```		
+<configuration>  
+  <connectionStrings>  
+		<add name="sagas" connectionStrings="--anAzureStorageConnectionString--"/>  
+	</connectionStrings> 
+</configuration>
+```
+4. what about conflicts?
+5. Run the following command: `NServiceBus.AzureStoragePersistence.SagaDeduplicator.exe directory=<directory> sagaTypeName=<sagaTypeName> sagaProperty=<sagaProperty> operation=Upload connectionString=<connectionString>`
+	- All of the commandline parameters, with the exception of `operation=Upload`, should be exactly the same as they were in step #3.
+
+
+## After the patch process
+Once the patch process has been completed there will be exceptions thrown, and logged, when duplicates are found in the course of normal message processing. The exception that is logged will be `DuplicateSagaFoundException` and will have the following message structure:
```

Comment:
@SimonCropp What's the best possible way to distinguish this situation then? My aim was to be as specific as possible, including the exception type. If you think this is important, it'll require change in the ASP.